stdlib/public/core/Integers.swift.gyb

```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
+    if rhs.leadingZeroBitCount < q.leadingZeroBitCount {
+      return (0, self)
     }
+    
+    // Calculate the number of bits before q and rhs line up,
+    // we can skip that many bits of iteration.
+    let initialOffset = q.leadingZeroBitCount + (DoubleWidth.bitWidth - rhs.leadingZeroBitCount) - 1
+
+    // Start with remainder capturing the high bits of q.
+    var r = q >> Magnitude(DoubleWidth.bitWidth - initialOffset)
+    q <<= Magnitude(initialOffset)
+
+    let highBit = ~(~0 >> 1) as Magnitude
+    for _ in initialOffset..<DoubleWidth.bitWidth {
+      r <<= 1
+      if q & highBit != 0 {
+        r += 1 as Magnitude
+      }
+      q <<= 1
 
-    var high: Low = 0
-    var low: Low = 0
-
-    func mkResult(_ isOverflow: Bool)
-      -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
-
-      // TODO: High() cast fails
-      let result = DoubleWidth<T>((High(high), low))
-      if isNegative {
-        return DoubleWidth<T>().subtractingReportingOverflow(result)
+      if r >= rhs {
+        q |= 1
+        r -= rhs
       }
-      return (partialValue: result, overflow: ArithmeticOverflow(isOverflow))
     }
 
-    var carry: Low = 0
-
-    let lhs = self.magnitude
-    let rhs = rhs.magnitude
-
-    // TODO: gyb me!
-    let a = mul(rhs._storage.low, lhs._storage.low, carry)
-    low += a.partial
-    carry = a.carry
-    /*_log("(II) 1 (\(high), \(low)) carry: \(carry)")*/
+    // Sign of remainder matches dividend
+    let remainder = self < 0
+      ? 0 - DoubleWidth(r)
+      : DoubleWidth(r)
 
-    let b = mul(rhs._storage.low, lhs._storage.high, carry)
-    high += b.partial
-    carry = b.carry
-    /*_log("(II) 2 (\(high), \(low)) carry: \(carry)")*/
-
-    if carry != 0 {
-      /*_log("(EE) overflow")*/
-      return mkResult(true)
+    if isNegative {
+      return (0 - DoubleWidth(q), remainder)
+    } else {
+      return (DoubleWidth(q), remainder)
     }
+  }
 
-    let c = mul(rhs._storage.high, lhs._storage.low, carry)
-    low += c.partial
-    carry = c.carry
-    /*_log("(II) 3 (\(high), \(low)) carry: \(carry)")*/
-
-    let d = mul(rhs._storage.high, lhs._storage.high, carry)
-    high += d.partial
-    carry = d.carry
-    /*_log("(II) 4 (\(high), \(low)) carry: \(carry)")*/
+  public func dividedReportingOverflow(by other: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    if other == 0 || (other < 0 && self == .min) {
+      return (self, .overflow)
+    }
 
-    /*if (carry > 0) { _log("(EE) overflow") }*/
-    return mkResult(carry > 0)
+    return (quotientAndRemainder(dividingBy: other).quotient, .none)
   }
 
-  public func dividedReportingOverflow(by other: DoubleWidth<T>)
-    -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
-    fatalError()
-  }
+  public func remainderReportingOverflow(dividingBy other: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    if other == 0 {
```

Comment:
This is funny, because the remainder (0) is representable, but the quotient is not.